### PR TITLE
Generate Typescript documentation from wit files

### PR DESCRIPTION
**Pull request by mjoerussell**
This PR uses the `jco types` command to generate Typescript definitions for the wasm crates instead of `jco transpile`. The reason is that `jco transpile` generates ts definitions by decoding WASM component binaries, whereas `jco types` parses .wit files directly and uses that information. This means that the `Resolve` instance created during `types` has access to the type documentation info, and the one created during `transpile` does not.

I've modified `infra check npm` with a new step which calls `jco types` and handles the resulting .d.ts files. When calling `jco transpile` we're now passing the option `--no-typescript` to avoid generating conflicting typescript outputs.

These changes depend on https://github.com/NomicFoundation/jco/pull/1, which augments the `types` command to accept the same configuration file that we pass to `transpile`. Without this, the codegen in `types` will not be able to generate Typescript enums for Rust enums, or any other customization that we want.

### CODE REVIEW

**Comment by mjoerussell** [comment]
The core changes are done, but I still need to debug an issue with the npm tests. After these changes Jest is not able to resolve the module imports, even though `infra check` still succeeds. I'll keep looking and see if there's a deeper issue.

File: Cargo.toml [hunk 1]
```diff
@@ -151,7 +151,8 @@ thiserror = { version = "1.0.68" }
 toml = { version = "0.8.19" }
 trybuild = { version = "1.0.101" }
 url = { version = "2.4.1", features = ["serde"] }
-wasm-tools = { version = "1.216.0" }
+wasm-tools = { version = "1.202.0" }
```
Comment (OmarTawfik) [inline Cargo.toml:154]: is this downgrade intended? not sure I understand the reason.
Also, nit blank lines.

File: Cargo.toml [hunk 2]
```diff
@@ -151,7 +151,8 @@ thiserror = { version = "1.0.68" }
...
+
 wit-bindgen = { version = "0.30.0" }
 
 [workspace.lints.rust]
```
Comment (OmarTawfik) [inline Cargo.toml:158]: for `submodules/jco`, I suggest updating it to: https://github.com/NomicFoundation/jco/commit/a951a946df8a2e4a9db3d12d210b925a93d7ccae

File: crates/infra/cli/src/toolchains/wasm/mod.rs [hunk 3]
```diff
@@ -92,10 +91,25 @@ impl WasmPackage {
                 .property("--out-dir", temp_dir.unwrap_str())
                 .property("--base64-cutoff", "0") // disable inlining core Wasm binaries as base64
                 .flag("--no-namespaced-exports") // disable namespaced exports for typescript compatibility
+                .flag("--no-typescript") // disable generating .d.ts files - we'll do this in `generate_types` with a better generator
```
Comment (OmarTawfik) [inline crates/infra/cli/src/toolchains/wasm/mod.rs:94]: nit: outdated comment?

```suggestion
                .flag("--no-typescript") // disable generating .d.ts files - we'll do this in `jco types` below
```

File: crates/infra/cli/src/toolchains/wasm/mod.rs [hunk 4]
```diff
@@ -110,7 +124,6 @@ impl WasmPackage {
                     let contents = temp_path.read_to_string()?;
                     fs.write_file(output_path, contents)?;
                 }
-
                 "js" => {
```
Comment (OmarTawfik) [inline crates/infra/cli/src/toolchains/wasm/mod.rs:127]: nit: left over?

File: crates/codegen/runtime/npm/package/wasm/generated/interfaces/nomic-foundation-slang-parser.d.ts [hunk 5]
```diff
@@ -15,21 +15,60 @@ import type { TextRange } from "./nomic-foundation-slang-cst.js";
 export { TextRange };
 
 export class ParseError {
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/wasm/generated/interfaces/nomic-foundation-slang-parser.d.ts:17]: I don't see the type-level documentation for many classes, like `ParseError` and `ParseOutput`. Am I missing something?

File: crates/codegen/runtime/npm/package/wasm/generated/interfaces/nomic-foundation-slang-cst.d.ts [hunk 6]
```diff
@@ -106,26 +294,62 @@ export class NonterminalNode {
   asTerminalNode(): undefined;
   isTerminalNode(): false;
```
Comment (OmarTawfik) [inline crates/codegen/runtime/npm/package/wasm/generated/interfaces/nomic-foundation-slang-cst.d.ts:295]: seems like these are also missing doc comments. But maybe that is more related to #1166 and can be done in a separate PR.

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Left a few questions.

File: crates/codegen/runtime/npm/package/wasm/generated/interfaces/nomic-foundation-slang-parser.d.ts (hunk 5 above)
Comment (OmarTawfik) [reply crates/codegen/runtime/npm/package/wasm/generated/interfaces/nomic-foundation-slang-parser.d.ts:17]: Sounds good. We can do that in #1171

**Review by OmarTawfik — APPROVED** [review]
I think there is still [one remaining issue](https://github.com/NomicFoundation/slang/pull/1157#discussion_r1861960929), but otherwise, LGTM.

**Pull request merged**